src/core/libmaven/PeakDetector.cpp

```diff
@@ -376,103 +527,435 @@ void PeakDetector::processSlices(vector<mzSlice*> &slices, string setName)
+                                       bool barplotIsotopes)
+{
+    if (_mavenParameters->allgroups.empty())
         return;
 
-    GroupFiltering groupFiltering(mavenParameters);
-    vector<PeakGroup> toBeMerged;
-    auto iter = mavenParameters->allgroups.begin();
-    while(iter != mavenParameters->allgroups.end()) {
-        auto& group = *iter;
-        bool matchFound = false;
-        for (auto compound : identificationSet) {
-            float mz = 0.0f;
-            if (compound->formula().length() || compound->neutralMass() != 0.0f) {
-                int charge = mavenParameters->getCharge(compound);
-                mz = compound->adjustedMass(charge);
+    sendBoostSignal("Performing meta-grouping…", 0, 0);
+
+    // separate parent groups, then filter for the N-best groups per compound
+    unordered_map<Compound*, vector<size_t>> parentCompounds;
+    for (size_t i = 0; i < _mavenParameters->allgroups.size(); ++i) {
+        if (_mavenParameters->stop)
+            return;
+
+        PeakGroup& group = _mavenParameters->allgroups[i];
+        Compound* compound = group.getCompound();
+        if (compound == nullptr)
+            continue;
+
+        if (group.isotope().isParent() && group.adduct()->isParent()) {
+            if (parentCompounds.count(compound) == 0)
+                parentCompounds[compound] = {};
+            parentCompounds[compound].push_back(i);
+        }
+    }
+    if (applyGroupFilters) {
+        _keepNBestRanked(parentCompounds,
+                         _mavenParameters->allgroups,
+                         _mavenParameters->eicMaxGroups);
+    }
+
+    // put isotopologues and adducts into separate buckets
+    unordered_map<Compound*, vector<size_t>> nonParentIsotopologues;
+    unordered_map<Compound*, vector<size_t>> nonParentAdducts;
+    for (size_t i = 0; i < _mavenParameters->allgroups.size(); ++i) {
+        if (_mavenParameters->stop)
+            return;
+
+        PeakGroup& group = _mavenParameters->allgroups[i];
+        Compound* compound = group.getCompound();
+        if (compound == nullptr)
+            continue;
+
+        if (group.isIsotope()) {
+            if (nonParentIsotopologues.count(compound) == 0)
+                nonParentIsotopologues[compound] = {};
+            nonParentIsotopologues[compound].push_back(i);
+        } else if (group.isAdduct()) {
+            if (nonParentAdducts.count(compound) == 0)
+                nonParentAdducts[compound] = {};
+            nonParentAdducts[compound].push_back(i);
+        }
+    }
+
+    // enumerate group IDs for all remaining peak-groups
+    int groupId = 1;
+    for (auto& group : _mavenParameters->allgroups)
+        group.setGroupId(groupId++);
+
+    if (nonParentIsotopologues.empty() && nonParentAdducts.empty())
+        return;
+
+    // lambda: given a compound and its child indexes, clubs them with their
+    // most likely parent-group if possible, otherwise adds them to a ghost
+    auto makeMeta =
```

Comment:
Could you move this out in a function instead? a 60-line lambda defined in a function hurts readability imo. 
Let me know if there's some advantage to this approach that I'm missing.

Reply:
It started out small. I do not think it hurts readability. In a good C++ editor, you can minimize these code-blocks. But most people will agree with you. I will move this out.